src/Symfony/Component/HttpFoundation/StreamedJsonResponse.php

```diff
@@ -0,0 +1,171 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpFoundation;
+
+/**
+ * StreamedJsonResponse represents a streamed HTTP response for JSON.
+ *
+ * A StreamedJsonResponse uses a structure and generics to create an
+ * efficient resource saving JSON response.
+ *
+ * It uses flush after a specified flush size to directly stream the data.
+ *
+ * @see flush()
+ *
+ * @author Alexander Schranz <alexander@sulu.io>
+ *
+ * Example usage:
+ *
+ * function loadArticles(): \Generator
+ *     // some streamed loading
+ *     yield ['title' => 'Article 1'];
+ *     yield ['title' => 'Article 2'];
+ *     yield ['title' => 'Article 3'];
+ * }),
+ *
+ * $response = new StreamedJsonResponse(
+ *     // json structure with generators in which will be streamed
+ *     [
+ *         '_embedded' => [
+ *             'articles' => loadArticles(), // any generator which you want to stream as list of data
+ *         ],
+ *     ],
+ * );
+ */
+class StreamedJsonResponse extends StreamedResponse
+{
+    public const DEFAULT_ENCODING_OPTIONS = JsonResponse::DEFAULT_ENCODING_OPTIONS;
+
+    private int $encodingOptions = self::DEFAULT_ENCODING_OPTIONS;
+
+    /**
+     * @param mixed[]                        $data      JSON Data containing PHP generators which will be streamed as list of data
+     * @param int                            $status    The HTTP status code (200 "OK" by default)
+     * @param array<string, string|string[]> $headers   An array of HTTP headers
+     * @param int                            $flushSize After every which item of a generator the flush function should be called
+     */
+    public function __construct(
+        private readonly array $data,
+        int $status = 200,
+        array $headers = [],
+        private int $flushSize = 100,
+    ) {
+        parent::__construct($this->stream(...), $status, $headers);
+
+        if (!$this->headers->get('Content-Type')) {
+            $this->headers->set('Content-Type', 'application/json');
+        }
+    }
+
+    private function stream(): void
+    {
+        $generators = [];
+        $structure = $this->data;
+
+        array_walk_recursive($structure, function (&$item, $key) use (&$generators) {
+            // generators should be used but for better DX all kind of Traversable are supported
+            if ($item instanceof \Traversable && !$item instanceof \JsonSerializable) {
+                // using uniqid to avoid conflict with eventually other data in the structure
+                $placeholder = uniqid('__placeholder_', true);
```

Comment:
It is possible to reuse the same placeholder for each generator:
- a sequential key in the `$generator` array is enough to map the generator to the each placeholder
- a single call to `uniqid` to better for performances
- a single call to `explode('"'.$placeholder'"', $structureText)` without limit will give all
- no need to `array_keys($generators)`

```php
// Unique placeholder
$placeholder = uniqid('__placeholder_', true);
array_walk_recursive($structure, function (&$item, $key) use (&$generators, $placeholder) {
    if ($item instanceof \Traversable && !$item instanceof \JsonSerializable) {
        // push to array to increment index
        $generators[] = $item;
        $item = $placeholder;
    }
}

// json encode

// Single split in an array of JSON parts
$parts = explode('"'.$placeholder'"', $structureText);
foreach($parts as $index => $part) {
    echo $part;
    if (!isset($generator[$index])) {
        // last part
        break;
    }
    foreach ($generator[$index] as $key => $item) {
        // ...
    }
    // ...
}
```

Reply:
I know there is low probability, but to prevent any random issue if `$item === $generator`, this could be checked and the string added to the `$generators` list.

```php
// Unique placeholder
$placeholder = uniqid('__placeholder_', true);
array_walk_recursive($structure, function (&$item, $key) use (&$generators, $placeholder) {
    if ($item instanceof \Traversable && !$item instanceof \JsonSerializable) {
        // push to array to increment index
        $generators[] = $item;
        $item = $placeholder;
    }
}

// json encode

// Single split in an array of JSON parts
$parts = explode('"'.$placeholder'"', $structureText);
foreach($parts as $index => $part) {
    echo $part;
    if (!isset($generator[$index])) {
        // last part
        break;
    }
    // Inject string if the placeholder conflicted with the value.
    if ($placeholder === $generator[$index]) {
        echo json_encode($generator[$index], $jsonEncodingOptions);
    } else {
        foreach ($generator[$index] as $key => $item) {
            // ...
        }
    }
    // ...
}
```

Reply:
Last thing. `uniqid` may not be necessary. Using a fixed value is enough once collusion issue is resolved.

`$placeholder = '__symfony_json__'`

Reply:
`uniqid` is required. If the structure json contains any "user" or database data someone can with `"__symfony_json__"` make that API crash. That is why I moved from `indexes` to `uniqid`.

Using "one" placeholder to split by that one and use then the index sounds like a good solution, like that idea.

Reply:
> `uniqid` is required. If the structure json contains any "user" or database data someone can with `"__symfony_json__"` make that API crash. That is why I moved from `indexes` to `uniqid`.

To avoid a conflict, the trick is to treat the string placeholder like the generators. 
```php
// in the array_walk_recursive
if ($placeholder === $item) {
    $generators[] = $item;
    $item = $placeholder;
}
```
And reinject the json encoded string when a $generator is a string
```php
// in the foreach
if ($placeholder === $generator[$index]) {
    echo json_encode($generator[$index], $jsonEncodingOptions);
} else // echo generator
```

Reply:
@alexander-schranz `uniqid` does not prevent conflicts. It only makes them random instead of reproducible.

Reply:
The implementation of @GromNaN is merged. So not longer a uniqid is required :)